src/slicegpt/model_adapter.py

```diff
@@ -293,6 +297,123 @@ def convert_layer_to_compressed_and_register_buffers(self, layer: Module, layer_
+        local_files_only: bool = False,
+        token: str | bool | None = None,
+    ) -> ModelAdapter:
+        """
+        Load the model from the given path and return the corresponding ModelAdapter instance.
+        `model_type` can be 'pretrained' or 'uninitialized'.
+        """
+
+        def find_recursively(adapter_cls: type[ModelAdapter]) -> ModelAdapter | None:
+            """
+            Recursively search for a subclass that can handle the model.
+            """
+            # depth first search to find the most specific subclass that can handle the model
+            for subclass in adapter_cls.__subclasses__():
+                candidate = find_recursively(subclass)
+                if candidate is not None:
+                    return candidate
+
+            if inspect.isabstract(adapter_cls):
+                return None
+
+            return adapter_cls._from_model(
+                model_name,
+                model_path=model_path,
+                model_type=model_type,
+                dtype=dtype,
+                local_files_only=local_files_only,
+                token=token,
+            )
+
+        adapter = find_recursively(cls)
+        if adapter is not None:
+            return adapter
+
+        raise NotImplementedError(f"{model_path} is neither a Hugging Face model nor a supported local model.")
+
+    @classmethod
+    def _from_model(
+        cls,
+        model_name: str,
+        model_path: str,
+        *,
+        model_type: str = 'pretrained',
+        dtype: torch.dtype = torch.float16,
+        local_files_only: bool = False,
+        token: str | bool | None = None,
+    ) -> ModelAdapter | None:
+        match model_type:
+            case 'pretrained':
+                return cls._from_pretrained(
+                    model_name,
+                    model_path=model_path,
+                    dtype=dtype,
+                    local_files_only=local_files_only,
+                    token=token,
+                )
+
+            case 'uninitialized':
+                return cls._from_uninitialised(
+                    model_name,
+                    model_path=model_path,
+                    dtype=dtype,
+                    local_files_only=local_files_only,
+                    token=token,
+                )
+            case _:
+                raise ValueError(f"Unknown model type: {model_type}")
+
+    @classmethod
+    @abstractmethod
+    def _from_pretrained(
+        cls,
+        model_name: str,
+        model_path: str,
+        *,
+        dtype: torch.dtype = torch.float16,
+        local_files_only: bool = False,
+        token: str | bool | None = None,
+    ) -> ModelAdapter | None:
+        """
+        Load the pretrained model from the given path and return a ModelAdapter instance.
+        Return None if the model_path is not supported.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    @abstractmethod
+    def _from_uninitialised(
```

Comment:
I suggest _from_uninitiali*z*ed, given that the rest of the codebase uses American spelling.